runtime/src/iree/hal/utils/wrapped_buffer.c

```diff
@@ -0,0 +1,152 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/hal/utils/wrapped_buffer.h"
+
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "iree/base/api.h"
+#include "iree/base/internal/synchronization.h"
+#include "iree/base/tracing.h"
+
+typedef struct iree_hal_wrapped_buffer_t {
+  iree_hal_buffer_t base;
+  iree_hal_buffer_t* wrapped_buffer;
+  iree_hal_buffer_params_t params;
+  iree_device_size_t allocation_size;
+  iree_allocator_t host_allocator;
+} iree_hal_wrapped_buffer_t;
+
+static const iree_hal_buffer_vtable_t iree_hal_wrapped_buffer_vtable;
+
+bool iree_hal_wrapped_buffer_isa(iree_hal_buffer_t* base_value) {
```

Comment:
This is unlikely to work in cases where dynamic linking is involved - we ran into issues with this approach previously and had to back it out. The vtable pointer in one library will not match the vtable pointer in another if they both have the vtable compiled in. That _may_ be ok for the extremely limited case of an internal type, but not for a generic type like this.